### PR TITLE
CSV Task - deliverable 4

### PR DESCRIPTION
**Pull request by tehreem-sadat**
Extend DB model entries and APIs to perform list and update for DataEXport Task

```
  Get the list of user tasks -> api/fx/export/v1/tasks/
  Patch on task -> api/fx/export/v1/tasks/<task_id>/
```

### CODE REVIEW

File: futurex_openedx_extensions/dashboard/serializers.py [hunk 1]
```diff
@@ -28,6 +30,27 @@
...
+        fields = '__all__'
+
+    def __init__(self, *args: Any, **kwargs: Any):
+        super().__init__(*args, **kwargs)
+        self.writable_fields = ['notes']
+        # Set fields to read-only unless they are in the writable list.
+        # DRF will ignore these values if included in the incoming data and only process writable fields.
+        for field_name in self.fields:
+            if field_name not in self.writable_fields:
+                self.fields[field_name].read_only = True
```
Comment (shadinaif) [inline futurex_openedx_extensions/dashboard/serializers.py:46]:
```suggestion
    class Meta:
        model = DataExportTask
        fields = '__all__'
        read_only_fields = [
            field.name for field in DataExportTask._meta.fields if field.name not in ['notes']
        ]
```

File: futurex_openedx_extensions/dashboard/views.py [hunk 2]
```diff
@@ -292,6 +293,34 @@ def get(self, request: Any, username: str, *args: Any, **kwargs: Any) -> JsonRes
...
+    ListAPIView, UpdateAPIView, FXViewRoleInfoMixin
+):  # pylint: disable=too-many-ancestors
+    """View to get the list of learners for a course"""
+    serializer_class = serializers.DataExportTaskSerializer
+    permission_classes = [FXHasTenantCourseAccess]
+    pagination_class = DefaultPagination
+    fx_view_name = 'exported_files_data'
+    fx_default_read_only_roles = ['staff', 'instructor', 'data_researcher', 'org_course_creator_group']
+    fx_default_read_write_roles = ['staff', 'instructor', 'data_researcher', 'org_course_creator_group']
+    fx_allowed_write_methods = ['PUT']
```
Comment (shadinaif) [inline futurex_openedx_extensions/dashboard/views.py:306]: `PATCH` is the right choice
```suggestion
    fx_allowed_write_methods = ['PATCH']
```

File: futurex_openedx_extensions/dashboard/views.py [hunk 3]
```diff
@@ -292,6 +293,34 @@ def get(self, request: Any, username: str, *args: Any, **kwargs: Any) -> JsonRes
...
+        """Get the list of user tasks"""
+        return DataExportTask.objects.filter(user=self.request.user)
+
+    def get_object(self) -> DataExportTask:
+        """Override to ensure that the user can only update their own tasks."""
+        task_id = self.kwargs.get('task_id')
+        task = get_object_or_404(DataExportTask, id=task_id, user=self.request.user)
+        return task
+
+
```
Comment (shadinaif) [inline futurex_openedx_extensions/dashboard/views.py:323]: disable `PUT` completely
```suggestion
    def put(self, request, *args, **kwargs):
        Response({"detail": "PUT method is not allowed."}, status=status.HTTP_405_METHOD_NOT_ALLOWED)


```

**Review by shadinaif — CHANGES_REQUESTED** [review]

File: futurex_openedx_extensions/helpers/tasks_utils.py [hunk 4]
```diff
@@ -97,15 +98,15 @@ def _paginated_response_generator(
         url = response.data.get('next')
 
 
-def _upload_file_to_storage(local_file_path: str, filename: str) -> str:
+def _upload_file_to_storage(local_file_path: str, filename: str, tenant: str) -> str:
```
Comment (shadinaif) [inline futurex_openedx_extensions/helpers/tasks_utils.py:101]:
```suggestion
def _upload_file_to_storage(local_file_path: str, filename: str, tenant_idt: int) -> str:
```

File: futurex_openedx_extensions/helpers/tasks_utils.py [hunk 5]
```diff
@@ -97,15 +98,15 @@ def _paginated_response_generator(
...
     """
     Upload a file to the default storage (e.g., S3).
 
     :param local_file_path: Path to the local file to upload
     :param filename: ilename for generated CSV
     :return: The path of the uploaded file
     """
-    storage_path = os.path.join(settings.FX_DATA_EXPORT_DIR_NAME, filename)
+    storage_path = os.path.join(settings.FX_DATA_EXPORT_DIR_NAME, tenant, filename)
```
Comment (shadinaif) [inline futurex_openedx_extensions/helpers/tasks_utils.py:109]: Sorry if it wasn't clear. Please change the default of `FX_DATA_EXPORT_DIR_NAME` to 

```python
    settings.FX_DATA_EXPORT_DIR_NAME = getattr(
        settings,
        'FX_DATA_EXPORT_DIR_NAME ',
        'exported_files'
    )
```

and create a function to return the directory, something like
```
def get_dir_of_storage(tenant_id: int, dir: str):
    return os.path.join(f'fx_dashboard/{tenant_id}', dir)
```

so here:
```suggestion
    storage_path = get_dir_of_storage(tenant_id, os.path.join(settings.FX_DATA_EXPORT_DIR_NAME, filename))
```

File: futurex_openedx_extensions/helpers/tasks_utils.py [hunk 6]
```diff
@@ -139,11 +141,13 @@ def _generate_csv_with_tracked_progress(
                     # update task progress
                     fx_task.progress = progress
                     fx_task.save()
-        storage_path = _upload_file_to_storage(tmp_file.name, filename)
+        tenant = TenantConfig.objects.get(id=fx_task.tenant_id)
+        storage_path = _upload_file_to_storage(tmp_file.name, filename, tenant.external_key)
```
Comment (shadinaif) [inline futurex_openedx_extensions/helpers/tasks_utils.py:145]: we always use `tenant.id`. The external key id for integration with external services
```suggestion
        storage_path = _upload_file_to_storage(tmp_file.name, filename, fx_task.tenant_id)
```

**Review by shadinaif — APPROVED** [review]
last request @tehreem-sadat , please squash and merge after applying the requested change

thank you for the amazing work!

File: futurex_openedx_extensions/helpers/tasks_utils.py (hunk 5 above)
Comment (tehreem-sadat) [reply futurex_openedx_extensions/helpers/tasks_utils.py]: done

**Pull request merged**